packages/material-ui/src/InputBase/InputBase.js

```diff
@@ -142,289 +141,259 @@ export const styles = theme => {
-    if (props.disabled && state.focused) {
-      return { focused: false };
-    }
-    return null;
+const InputBase = React.forwardRef(function InputBase(props, ref) {
+  const {
+    autoComplete,
+    autoFocus,
+    classes,
+    className: classNameProp,
+    defaultValue,
+    disabled,
+    endAdornment,
+    error,
+    fullWidth,
+    id,
+    inputComponent,
+    inputProps: { className: inputPropsClassName, ...inputPropsProp } = {},
+    inputRef: inputRefProp,
+    margin,
+    muiFormControl,
+    multiline,
+    name,
+    onBlur,
+    onChange,
+    onClick,
+    onEmpty,
+    onFilled,
+    onFocus,
+    onKeyDown,
+    onKeyUp,
+    placeholder,
+    readOnly,
+    renderPrefix,
+    rows,
+    rowsMax,
+    startAdornment,
+    type,
+    value,
+    ...other
+  } = props;
+
+  const inputRef = React.useRef();
+  const { current: isControlled } = React.useRef(props.value != null);
+  const [focused, setFocused] = React.useState(false);
+
+  const handleInputRefWarning = instance => {
+    warning(
+      !instance || instance instanceof HTMLInputElement || instance.focus,
+      [
+        'Material-UI: you have provided a `inputComponent` to the input component',
+        'that does not correctly handle the `inputRef` property.',
+        'Make sure the `inputRef` property is called with a HTMLInputElement.',
+      ].join('\n'),
+    );
+  };
+  const handleInputPropsRefProp = useForkRef(inputPropsProp.ref, handleInputRefWarning);
+  const handleInputRefProp = useForkRef(inputRefProp, handleInputPropsRefProp);
+  const handleRef = useForkRef(inputRef, handleInputRefProp);
+
+  // The blur won't fire when the disabled state is set on a focused input.
+  // We need to book keep the focused state manually.
+  if (disabled && focused) {
+    setFocused(false);
   }
 
-  constructor(props) {
-    super(props);
-    this.isControlled = props.value != null;
-  }
+  const checkDirty = obj => {
+    if (isFilled(obj)) {
+      if (muiFormControl && muiFormControl.onFilled) {
+        muiFormControl.onFilled();
+      }
+      if (onFilled) {
+        onFilled();
+      }
+      return;
+    }
 
-  state = {
-    focused: false,
+    if (muiFormControl && muiFormControl.onEmpty) {
+      muiFormControl.onEmpty();
+    }
+    if (onEmpty) {
+      onEmpty();
+    }
   };
 
-  componentDidMount() {
-    this.checkDirty(this.isControlled ? this.props : this.inputRef);
-  }
-
-  componentDidUpdate(prevProps) {
+  React.useEffect(() => {
     // Book keep the focused state.
-    if (!prevProps.disabled && this.props.disabled) {
-      const { muiFormControl } = this.props;
+    if (disabled) {
       if (muiFormControl && muiFormControl.onBlur) {
         muiFormControl.onBlur();
       }
     }
-    if (this.isControlled) {
-      this.checkDirty(this.props);
-    } // else performed in the onChange
-  }
 
-  handleFocus = event => {
-    const { muiFormControl } = this.props;
+    checkDirty(isControlled ? props : inputRef.current);
```

Comment:
This means that when you change the onEmpty prop it gets called instantly if it's empty. I'm not sure if this is intended behaviour.

Reply:
The initial code was like this;

```diff
-componentDidMount() {
-  this.checkDirty(this.isControlled ? this.props : this.inputRef);
-}

-componentDidUpdate(prevProps) {
-  if (this.isControlled) {
-     this.checkDirty(this.props);
-  }
-}
```

I converted that into 

```js
React.useEffect(() => {
   checkDirty(isControlled ? props : inputRef.current);
})
```

Did I do it wrong here?

Reply:
Hmm, it reflects what was there already I guess.

Reply:
So how do you propose I should move with this particular test? 💭

Reply:
Well, it's just changing the callCount. @oliviertassinari does this seem right to you?

Reply:
It looks like uncontrolled shouldn't checkDirty. 
```js
React.useEffect(() => {
  if (!isControlled) checkDiry(this.inputRef)
}, [])
React.useEffect(() => {
  if (isControlled) checkDiry(props)
});
```
is closer to the original implementation. When in doubt stick with the old code.

Reply:
I have made the changes, can you kindly review the PR. @joshwooding @eps1lon :)